### PR TITLE
[FLINK-23652][datastream] FLIP-179: Add common connector metrics

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Currently, Flink exposes a generic MetricGroup to UDFs through RuntimeContext , such that users can implement their own custom metrics. However, without some kind of standardization, the custom metrics are unnecessarily hard to monitor. Metrics for sources and sinks are especially valuable and thus the effort to standardize the names in [FLIP-33](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics) was started.

[FLIP-179](https://cwiki.apache.org/confluence/display/FLINK/FLIP-179%3A+Expose+Standardized+Operator+Metrics) draws upon the ideas of FLIP-33 but extends it for all UDFs. In the process, some changes to FLIP-33 are necessary to better align with the generalization.

Another motivation for this PR lies in the way the metric system works: The metric system does not allow fetching pre-created metrics by name; you either have to lazily create it or supply a getter to the metric. Thus, we need to add a collection of interfaces that properly expose the standardized metrics. Note that this FLIP covers only those metrics (coming from FLIP-33) that cannot be automatically calculated as only these metrics need to be exposed.

## Brief change log

- Extract common context creation in CollectionExecutor
- Extract Operator(IO)MetricGroup interfaces and expose them in RuntimeContext
- Adding test reporter and using it by default in MiniClusterResource to implement ITCases for metrics.
- Adding common source+sink metrics

## Verifying this change

Added unit and integration tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                    () -> clock.absoluteTimeMillis() - watermarkGauge.getValue());
+        } else {
+            watermarkGauge.setValue(watermark);
+        }
+    }
+
+    public void eventTimeEmitted(long timestamp) {
+        // iff a respective source emits a timestamp, Flink can provide the event lag
+        if (eventTimeGauge == null) {
+            eventTimeGauge = new SettableGauge<>(timestamp);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java:112]: What exactly does the SettableGauge provide us here opposed to a plain `long` that defaults to -1?

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/SettableGauge.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.metrics;
+
+import java.util.Objects;
+
+/** A gauge that returns the last set value. */
+public class SettableGauge<T> implements Gauge<T> {
```
Comment (zentol) [inline flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/SettableGauge.java:23]: I still don't quite see a need for this class.

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/OperatorIOMetricGroup.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.MetricGroup;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Metric group that contains shareable pre-defined IO-related metrics for operators.
+ *
+ * <p>All metrics can only be accessed in the main operator thread.
```
Comment (zentol) [inline flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/OperatorIOMetricGroup.java:28]: This is a bit misleading for internal usages because we do access metrics outside the operator thread, from the reporter thread.
I also wonder whether "can" is the right word; you "can" totally access them from a different thread. You _shouldn't_ though (in general).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 4]
```diff
@@ -619,9 +618,7 @@ public StreamTaskStateInitializer createStreamTaskStateInitializer() {
 
     protected Counter setupNumRecordsInCounter(StreamOperator streamOperator) {
         try {
-            return ((OperatorMetricGroup) streamOperator.getMetricGroup())
-                    .getIOMetricGroup()
-                    .getNumRecordsInCounter();
+            return streamOperator.getMetricGroup().getIOMetricGroup().getNumRecordsInCounter();
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:621]: 🎉

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/reporter/TestReporter.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public void close() {
+        metrics.clear();
+        REPORTERS.remove();
+    }
+
+    public static TestReporter getInstance() {
+        return REPORTERS.get();
+    }
```
Comment (zentol) [inline flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/reporter/TestReporter.java:47]: nit: move this to the top

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/OperatorIOMetricGroup.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ */
+@NotThreadSafe
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/OperatorIOMetricGroup.java:30]: You are right. I'm removing it. Definition of ThreadSafe

```
 * The class to which this annotation is applied is thread-safe. This means that
 * no sequences of accesses (reads and writes to public fields, calls to public
 * methods) may put the object into an invalid state, regardless of the
 * interleaving of those actions by the runtime, and without requiring any
 * additional synchronization or coordination on the part of the caller.
 ```

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/OperatorIOMetricGroup.java (hunk 3 above)
Comment (AHeise) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/OperatorIOMetricGroup.java:28]: I replaced with `You should only update the metrics in the main operator thread.`

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java (hunk 1 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java]: Gotcha.

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/SettableGauge.java (hunk 2 above)
Comment (AHeise) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/SettableGauge.java]: Removed.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/mocks/MockRecordEmitter.java [hunk 7]
```diff
@@ -20,20 +20,46 @@
...
 
 /**
  * A mock {@link RecordEmitter} that works with the {@link MockSplitReader} and {@link
  * MockSourceReader}.
  */
-public class MockRecordEmitter implements RecordEmitter<int[], Integer, AtomicInteger> {
+public class MockRecordEmitter implements RecordEmitter<int[], Integer, MockSplitState> {
+    public static final int RECORD_SIZE_IN_BYTES = 10;
+    private final SourceReaderMetricGroup metricGroup;
+    private final Map<MockSplitState, Boolean> knownSplits = new IdentityHashMap<>();
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/mocks/MockRecordEmitter.java:35]:
```suggestion
    private final Set<MockSplitState> knownSplits = Collections.newSetFromMap(new IdentityHashMap<>());
```

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/mocks/MockSplitState.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.base.source.reader.mocks;
+
+/** The state of the {@link MockSplitReader} reflecting the progress. */
+class MockSplitState {
+    private int index;
+    private final int endIndex;
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/mocks/MockSplitState.java:23]: maybe rename these to *RecordIndex?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 9]
```diff
@@ -19,25 +19,166 @@
...
+import static org.hamcrest.Matchers.both;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.lessThan;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
 
 /** IT case for the {@link Source} with a coordinator. */
 public class CoordinatedSourceITCase extends AbstractTestBase {
+    @Rule public SharedObjects sharedObjects = SharedObjects.create();
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:64]:
```suggestion
    @Rule public final SharedObjects sharedObjects = SharedObjects.create();
```

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 10]
```diff
@@ -19,25 +19,166 @@
...
+        afterBarrier.get().await();
+
+        beforeBarrier.get().await();
+        assertSourceMetrics(10, numRecordsPerSplit, env.getParallelism());
+        afterBarrier.get().await();
+
+        jobClient.getJobExecutionResult().get();
+    }
+
+    private void assertSourceMetrics(long processedRecords, long numTotal, int parallelism) {
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:121]:
```suggestion
    private void assertSourceMetrics(long processedRecordsPerSubtask, long numTotalPerSubtask, int parallelism) {
```

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 11]
```diff
@@ -19,25 +19,166 @@
...
+                                });
+        stream.addSink(new DiscardingSink<>());
+        JobClient jobClient = env.executeAsync();
+
+        beforeBarrier.get().await();
+        assertSourceMetrics(4, numRecordsPerSplit, env.getParallelism());
+        afterBarrier.get().await();
+
+        beforeBarrier.get().await();
+        assertSourceMetrics(10, numRecordsPerSplit, env.getParallelism());
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:115]:
```suggestion
        assertSourceMetrics(numRecordsPerSplit, numRecordsPerSplit, env.getParallelism());
```

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 12]
```diff
@@ -19,25 +19,166 @@
...
+                                        beforeBarrier.get().await();
+                                        afterBarrier.get().await();
+                                    }
+                                    return i;
+                                });
+        stream.addSink(new DiscardingSink<>());
+        JobClient jobClient = env.executeAsync();
+
+        beforeBarrier.get().await();
+        assertSourceMetrics(4, numRecordsPerSplit, env.getParallelism());
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:111]: replace magic number

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 13]
```diff
@@ -19,25 +19,166 @@
...
+    private static final long EVENTTIME_LAG = Duration.ofDays(100).toMillis();
+    private static final long WATERMARK_LAG = Duration.ofDays(1).toMillis();
+    private static final long EVENTTIME_EPSILON = Duration.ofDays(20).toMillis();
+    // this basically is the time a build is allowed to be frozen before the test fails
+    private static final long WATERMARK_EPSILON = Duration.ofHours(6).toMillis();
+
+    @Test
+    public void testMetrics() throws Exception {
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(4);
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:76]: add a comment that this should remain higher than the number of splits

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 14]
```diff
@@ -19,25 +19,166 @@
...
+            assertThat(
+                    metrics.get(MetricNames.NUM_RECORDS_IN_ERRORS),
+                    isCounter(equalTo(processedRecords / 2)));
+            // Timestamp assigner subtracting EVENTTIME_LAG from wall clock, so expect that lag
+            assertThat(
+                    metrics.get(MetricNames.CURRENT_EMIT_EVENT_TIME_LAG),
+                    isGauge(
+                            both(greaterThan(EVENTTIME_LAG - EVENTTIME_EPSILON))
+                                    .and(lessThan(EVENTTIME_LAG + EVENTTIME_EPSILON))));
+            // Watermark is derived from timestamp, so it has to be in the same OOM
```
Comment (zentol) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:155]: OOM?

Comment (AHeise) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:155]: order of magnitude - I replaced it

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalOperatorIOMetricGroup.java [hunk 15]
```diff
@@ -21,21 +21,26 @@
...
 import org.apache.flink.metrics.MeterView;
+import org.apache.flink.metrics.groups.OperatorIOMetricGroup;
 import org.apache.flink.runtime.metrics.MetricNames;
 
 /**
  * Metric group that contains shareable pre-defined IO-related metrics. The metrics registration is
  * forwarded to the parent operator metric group.
  */
-public class OperatorIOMetricGroup extends ProxyMetricGroup<OperatorMetricGroup> {
+public class InternalOperatorIOMetricGroup extends ProxyMetricGroup<InternalOperatorMetricGroup>
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalOperatorIOMetricGroup.java:31]: `@Internal`?

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/metrics/testutils/MetricMatchers.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+import java.util.function.Function;
+
+/**
+ * Provides hamcrest Matchers for metric assertions.
+ *
+ * @see org.apache.flink.metrics.reporter.TestReporter
+ */
+public enum MetricMatchers {
+    ;
```
Comment (fapaul) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/metrics/testutils/MetricMatchers.java:37]: Nit: Remove `;`?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/SourceReaderBaseTest.java [hunk 17]
```diff
@@ -143,7 +143,11 @@ public void testMultipleSplitsWithDifferentFinishingMoments() throws Exception {
                         .setBlockingFetch(false)
                         .build();
         MockSourceReader reader =
-                new MockSourceReader(elementsQueue, () -> mockSplitReader, getConfig(), null);
+                new MockSourceReader(
+                        elementsQueue,
+                        () -> mockSplitReader,
+                        getConfig(),
+                        new TestingReaderContext());
```
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/SourceReaderBaseTest.java:150]: Why did you change it from `null` to a value?

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    @VisibleForTesting
+    public static InternalSourceReaderMetricGroup mock(MetricGroup metricGroup) {
+        return new InternalSourceReaderMetricGroup(
+                metricGroup,
+                UnregisteredMetricsGroup.createUnregisteredOperatorIOMetricGroup(),
+                SystemClock.getInstance());
+    }
+
+    private boolean isIdling() {
+        return idleStartTime != ACTIVE;
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java:78]: It seems strange that if the source reader was idle once the idleTime will always increase. I would like to reset it somehow to 0.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 19]
```diff
@@ -19,30 +19,175 @@
...
+                    group.getIOMetricGroup().getNumBytesInCounter(),
+                    isCounter(
+                            equalTo(
+                                    processedRecordsPerSubtask
+                                            * MockRecordEmitter.RECORD_SIZE_IN_BYTES)));
+            // MockRecordEmitter is just incrementing errors every even record
+            assertThat(
+                    metrics.get(MetricNames.NUM_RECORDS_IN_ERRORS),
+                    isCounter(equalTo(processedRecordsPerSubtask / 2)));
+            // Timestamp assigner subtracting EVENTTIME_LAG from wall clock, so expect that lag
```
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:158]: Comment incomplete?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 20]
```diff
@@ -19,30 +19,175 @@
...
+            assertThat(
+                    metrics.get(MetricNames.CURRENT_EMIT_EVENT_TIME_LAG),
+                    isGauge(
+                            both(greaterThan(EVENTTIME_LAG - EVENTTIME_EPSILON))
+                                    .and(lessThan(EVENTTIME_LAG + EVENTTIME_EPSILON))));
```
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:163]: Ni: extract a function for this functionality

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/SinkWriterMetricGroup.java [hunk 21]
```diff
@@ -41,5 +41,5 @@
      * For example, a `volatile` field could be set in the async thread and lazily read in the
      * gauge.
      */
-    <G extends Gauge<Long>> G addCurrentSendTimeGauge(G currentSendTime);
+    void setCurrentSendTimeGauge(Gauge<Long> currentSendTimeGauge);
```
Comment (fapaul) [inline flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/groups/SinkWriterMetricGroup.java:44]: I do not know if the signature changes belong to this commit. IMO they should be part of the core commit.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java [hunk 22]
```diff
@@ -256,6 +275,95 @@ public void writerAndGlobalCommitterExecuteInBatchMode() throws Exception {
                 containsInAnyOrder(EXPECTED_GLOBAL_COMMITTED_DATA_IN_BATCH_MODE.toArray()));
     }
 
+    @Test
+    public void testMetrics() throws Exception {
```
Comment (fapaul) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java:279]: The code for the source metric tests looks very similar can you try to generify this a bit?

**Review by fapaul — COMMENTED** [review]
Thanks for the effort, overall the structure looks already pretty. I left a couple of nitty things but the idle time definition looks weird to me currently. PTAL

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/reporter/TestReporter.java [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A mock implementation that makes all reported metrics available for tests. Metrics remain
+ * registered even after the job finishes for easier assertions.
+ */
+public class TestReporter implements MetricReporter {
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/reporter/TestReporter.java:44]: I'll try to move it into the test-jar of the module.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/SourceReaderBaseTest.java (hunk 17 above)
Comment (AHeise) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/SourceReaderBaseTest.java:150]: Because the `SourceReaderBase` is now accessing the metric group in ctor.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java (hunk 19 above)
Comment (AHeise) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:158]: Probably just bad english; I'll try to fix.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java (hunk 22 above)
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java:279]: The are in completely unrelated packages and I hate to generify it to something completely independent. Maybe we should embed it into the connector testing framework when it arrives and is beefed for sinks.

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/reporter/TestReporter.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/reporter/TestReporter.java]: Moved it to test-jar of runtime where MiniClusterResource resides.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricNames.java [hunk 24]
```diff
@@ -75,4 +75,17 @@ public static String currentInputWatermarkName(int index) {
...
+    public static final String NUM_RECORDS_IN_ERRORS = "numRecordsInErrors";
+    public static final String CURRENT_FETCH_EVENT_TIME_LAG = "currentFetchEventTimeLag";
+    public static final String CURRENT_EMIT_EVENT_TIME_LAG = "currentEmitEventTimeLag";
+    public static final String WATERMARK_LAG = "watermarkLag";
+    public static final String PENDING_RECORDS = "pendingRecords";
+    public static final String PENDING_BYTES = "pendingBytes";
+    public static final String SOURCE_IDLE_TIME = "sourceIdleTime";
+
+    // FLIP-33 sink
+    public static final String NUM_RECORDS_OUT_ERRORS = "numRecordsOutErrors";
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricNames.java:89]: Nit: the commit only states something about source metrics ;)

**Review by fapaul — APPROVED** [review]
LGTM 👍  Really great effort :shipit:

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricNames.java (hunk 24 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricNames.java:89]: Thanks, I will separate.

**Comment by AHeise** [comment]
@flinkbot run azure

File: flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java [hunk 25]
```diff
@@ -256,6 +278,95 @@ public void writerAndGlobalCommitterExecuteInBatchMode() throws Exception {
...
+        afterBarrier.get().await();
+
+        jobClient.getJobExecutionResult().get();
+    }
+
+    private void assertSinkMetrics(
+            long processedRecordsPerSubtask, int parallelism, int numSplits) {
+        List<OperatorMetricGroup> groups =
+                inMemoryReporter.getReporter().findOperatorMetricGroups("TestSink");
+        //        assertThat(groups, hasSize(parallelism));
```
Comment (tillrohrmann) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java:330]: Can this statement be deleted?

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * A {@link MetricReporter} implementation that makes all reported metrics available for tests.
+ * Metrics remain registered even after the job finishes for easier assertions.
+ *
+ * <p>This class can only be accessed through {@link InMemoryReporterRule#getReporter()} to ensure
+ * that test cases are properly isolated.
+ */
+@Experimental
+public class InMemoryReporter implements MetricReporter {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java:55]: I think this class will be accessed from different threads but is not thread safe. This is a problem as seen in https://issues.apache.org/jira/browse/FLINK-23785.

**Review by tillrohrmann — COMMENTED** [review]
I think the `InMemoryReporter` needs to be thread safe to work.

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java (hunk 26 above)
Comment (AHeise) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java:55]: `MetricReporter`s are usually synchronized externally but I used it in an unintended way, so you are right it needs synchronization but for more elaborate reason. And it's not solving the whole issue, see the ticket.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java (hunk 25 above)
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkITCase.java:330]: No but I re-enabled it.

**Comment by pnowojski** [comment]
"The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)"

FYI, that wasn't true: https://issues.apache.org/jira/browse/FLINK-23776

Please be more careful next time.
